controllers/certificatesigningrequest_controller.go

```diff
@@ -0,0 +1,93 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	capi "k8s.io/api/certificates/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/cert-manager/signer/api"
+	capihelper "github.com/cert-manager/signer/internal/kubernetes/apis/certificates/v1beta1"
+	"github.com/cert-manager/signer/internal/kubernetes/signer"
+)
+
+// CertificateSigningRequestSigningReconciler reconciles a CertificateSigningRequest object
+type CertificateSigningRequestSigningReconciler struct {
+	client.Client
+	Log           logr.Logger
+	Scheme        *runtime.Scheme
+	SignerName    string
+	Signer        *signer.Signer
+	EventRecorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=certificates.k8s.io.cert-manager.io,resources=certificatesigningrequests,verbs=get;list;watch
+// +kubebuilder:rbac:groups=certificates.k8s.io.cert-manager.io,resources=certificatesigningrequests/status,verbs=patch
+
+func (r *CertificateSigningRequestSigningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("certificatesigningrequest", req.NamespacedName)
+	var csr capi.CertificateSigningRequest
+	if err := r.Client.Get(ctx, req.NamespacedName, &csr); client.IgnoreNotFound(err) != nil {
+		return ctrl.Result{}, fmt.Errorf("error %q getting CSR", err)
+	}
+	switch {
+	case !csr.DeletionTimestamp.IsZero():
+		log.Info("CSR has been deleted. Ignoring.")
+	case csr.Spec.SignerName == nil:
+		log.Info("CSR does not have a signer name. Ignoring.")
+	case *csr.Spec.SignerName != r.SignerName:
+		log.Info("CSR signer name does not match. Ignoring.", "signer-name", csr.Spec.SignerName)
+	case csr.Status.Certificate != nil:
+		log.Info("CSR has already been signed. Ignoring.")
+	case !api.IsCertificateRequestApproved(&csr):
+		log.Info("CSR is not approved, Ignoring.")
+	default:
+		log.Info("Signing")
+		x509cr, err := capihelper.ParseCSR(csr.Spec.Request)
+		if err != nil {
+			log.Error(err, "unable to parse csr")
+			r.EventRecorder.Event(&csr, v1.EventTypeWarning, "SigningFailed", "Unable to parse the CSR request")
+			return ctrl.Result{}, nil
+		}
+		cert, err := r.Signer.Sign(x509cr, csr.Spec.Usages)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error auto signing csr: %v", err)
```

Comment:
Will the `Sign` function ever error due to 'invalid input' or will it exclusively error only if an actual error occurs? Else we may enter a loop here 😄

Reply:
It will error if the CA file is invalid, and if the CSR is invalid, or it fails to get random data when it generates the certificate, plus some other errors that I've probably missed.

The invalid CSR is already checked earlier in the code and does not cause a reconcile loop.
The invalid CA file is something that may cause a reconcile loop, but that might be the correct behaviour, if for example, the file  is eventually fixed by https://github.com/jetstack/cert-manager-csi auto-renewal. (although I've never used that project and might be misunderstanding how it works.)